lgc/state/PalMetadata.cpp

```diff
@@ -969,3 +920,178 @@ void PalMetadata::eraseFragmentInputInfo() {
+  constexpr unsigned numberOfShaderStages = 5;
+  static const std::pair<unsigned, unsigned> shaderTable[numberOfShaderStages] = {
+      {mmSPI_SHADER_PGM_RSRC1_LS, CallingConv::AMDGPU_LS},
+      {mmSPI_SHADER_PGM_RSRC1_HS, CallingConv::AMDGPU_HS},
+      {mmSPI_SHADER_PGM_RSRC1_ES, CallingConv::AMDGPU_ES},
+      {mmSPI_SHADER_PGM_RSRC1_GS, CallingConv::AMDGPU_GS},
+      {mmSPI_SHADER_PGM_RSRC1_VS, CallingConv::AMDGPU_VS}};
+
+  for (unsigned i = 0; i < numberOfShaderStages; ++i) {
+    auto entry = m_registers.find(m_document->getNode(shaderTable[i].first));
+    if (entry != m_registers.end())
+      return shaderTable[i].second;
+  }
+  return CallingConv::AMDGPU_CS;
+}
+
+// =====================================================================================================================
+// Returns the offset of the first user data register for the given calling convention.
+//
+// @param callingConv : The calling convention
+unsigned PalMetadata::getFirstUserDataReg(unsigned callingConv) {
+  static const std::map<unsigned, unsigned> shaderTable = {{CallingConv::AMDGPU_LS, mmSPI_SHADER_USER_DATA_LS_0},
+                                                           {CallingConv::AMDGPU_HS, mmSPI_SHADER_USER_DATA_HS_0},
+                                                           {CallingConv::AMDGPU_ES, mmSPI_SHADER_USER_DATA_ES_0},
+                                                           {CallingConv::AMDGPU_GS, mmSPI_SHADER_USER_DATA_GS_0},
+                                                           {CallingConv::AMDGPU_VS, mmSPI_SHADER_USER_DATA_VS_0}};
+  static const std::map<unsigned, unsigned> shaderTableGfx9 = {{CallingConv::AMDGPU_LS, mmSPI_SHADER_USER_DATA_LS_0},
+                                                               {CallingConv::AMDGPU_HS, mmSPI_SHADER_USER_DATA_LS_0},
+                                                               {CallingConv::AMDGPU_ES, mmSPI_SHADER_USER_DATA_ES_0},
+                                                               {CallingConv::AMDGPU_GS, mmSPI_SHADER_USER_DATA_ES_0},
+                                                               {CallingConv::AMDGPU_VS, mmSPI_SHADER_USER_DATA_VS_0}};
+
+  bool isGfx9 = m_pipelineState->getTargetInfo().getGfxIpVersion().major == 9;
+  const std::map<unsigned, unsigned> &currentShaderTable = (isGfx9 ? shaderTableGfx9 : shaderTable);
+  return currentShaderTable.at(callingConv);
+}
+
+// =====================================================================================================================
+// Returns to number of SGPRs before the SGPRs used for other purposes before the first user data SGPRs.
+//
+// @param conv : The calling convention of the shader
+unsigned PalMetadata::getNumberOfSgprsBeforeUserData(unsigned conv) {
+  // GFX9+ merged shader have an extra 8 SGPRs before user data.
+  if (conv != CallingConv::AMDGPU_VS && m_pipelineState->getTargetInfo().getGfxIpVersion() >= GfxIpVersion{9, 0, 0})
```

Comment:
As this code now claims to be general and someone else might use it, perhaps it should work for CS (for which the answer should be 0), or at least assert that it is not being asked about CS.